### DISCUSSION TITLE
How do I dependency inject Sentry?

### DISCUSSION THREAD

**Discussion by ajacques** [getsentry/sentry-java#3047] [Q&A] [answered]
I'm working on upgrading from Sentry Java 1.x to Sentry Java 6.x and I've been using dependency inject to inject my SentryClient into classes. In the past, it was easy. `Sentry.init` returned a `SentryClient` and I passed that around my Guice/Dagger context.

Now, `Sentry.init` returns an `IHub` which contains the same methods as `ISentryClient` and more. But the docs recommend to use `ISentryClient` when you only need to report events and pass `IHub` when you need to configure the client (add tags, etc.)

There appears to be no public way to get an `ISentryClient` from the response of `Sentry.init`, nor does there appear to be a way to initialize a `Hub` with a specific `SentryClient` without actually initializing a `SentryClient` in the `Hub` constructor.

What is the recommended way to dependency inject a SentryClient?

**Comment by adinauer** [comment, marked as answer]
Hello @ajacques how about injecting `IHub` instances into your classes? You could use `HubAdapter` to forward to `Sentry` static method calls in case this is just for making your application more easily testable and mocking Sentry.

Are you planning on using multiple Sentry projects / DSNs to send data to from a single app? If no, there should be no need to deal with clients directly.

> the docs recommend to use ISentryClient when you only need to report events and pass IHub when you need to configure the client (add tags, etc.)

Can you please link the docs page?

**Reply by ajacques** [reply]
I can inject `IHub` instances since I only intend to have one target at runtime. Only during unit testing do I want to be able to inject different clients to verify errors are reported, but it's still one at a time.

> Can you please link the docs page?

Maybe it wasn't explicitly stated, but I came to this conclusion after asking the Sentry Chat bot when I should use one vs the other, and I saw [this](https://docs.sentry.io/platforms/dotnet/guides/aspnetcore/unit-testing/) which lead me to the idea that I should selectively support injecting one or the other.

**Reply by adinauer** [reply]
You somehow ended up in the dotnet docs. We should document how to do unit testing for the Java SDK. I've created https://github.com/getsentry/sentry-java/issues/3050 to track.
